packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts

```diff
@@ -78,135 +31,190 @@ export const useGridKeyboardNavigation = (
-          const viewportPageSize = apiRef.current.unstable_getViewportPageSize();
-          const field = apiRef.current.getVisibleColumns()[colIndex].field;
-          const id = currentPage.rows[Math.min(viewportPageSize, currentPage.rows.length - 1)].id;
-          apiRef.current.setCellFocus(id, field);
+      const viewportPageSize = apiRef.current.unstable_getViewportPageSize();
+      const colIndexBefore = params.field ? apiRef.current.getColumnIndex(params.field) : 0;
+      const firstRowIndexInPage = currentPage.range?.firstRowIndex ?? null;
+      const lastRowIndexInPage = currentPage.range?.lastRowIndex ?? null;
+      const firstColIndex = 0;
+      const lastColIndex = colCount - 1;
+
+      // eslint-disable-next-line default-case
+      switch (event.key) {
+        case 'ArrowDown':
+        case 'Enter': {
+          if (firstRowIndexInPage !== null) {
+            goToCell(colIndexBefore, firstRowIndexInPage);
+          }
+          break;
         }
-        return;
-      } else {
-        throw new Error('MUI: Key not mapped to navigation behavior.');
-      }
 
-      if (!nextColumnHeaderIndexes) {
-        const field = apiRef.current.getVisibleColumns()[colIndex].field;
-        if (currentPage.rows.length) {
-          apiRef.current.setCellFocus(currentPage.rows[0].id, field);
+        case 'ArrowRight': {
+          if (colIndexBefore < lastColIndex) {
+            goToHeader(colIndexBefore + 1, event);
+          }
+          break;
         }
-        return;
-      }
 
-      nextColumnHeaderIndexes!.colIndex = Math.max(0, nextColumnHeaderIndexes!.colIndex);
-      nextColumnHeaderIndexes!.colIndex =
-        nextColumnHeaderIndexes!.colIndex >= colCount
-          ? colCount - 1
-          : nextColumnHeaderIndexes!.colIndex;
+        case 'ArrowLeft': {
+          if (colIndexBefore > firstColIndex) {
+            goToHeader(colIndexBefore - 1, event);
+          }
+          break;
+        }
 
-      logger.debug(`Navigating to next column row ${nextColumnHeaderIndexes.colIndex}`);
-      apiRef.current.scrollToIndexes(nextColumnHeaderIndexes);
-      const field = apiRef.current.getVisibleColumns()[nextColumnHeaderIndexes.colIndex].field;
-      apiRef.current.setColumnHeaderFocus(field, event);
+        case 'PageDown': {
+          if (firstRowIndexInPage !== null && lastRowIndexInPage !== null) {
+            goToCell(
+              colIndexBefore,
+              Math.min(firstRowIndexInPage + viewportPageSize, lastRowIndexInPage),
+            );
+          }
+          break;
+        }
+
+        case 'Home': {
+          goToHeader(firstColIndex, event);
+          break;
+        }
+
+        case 'End': {
+          goToHeader(lastColIndex, event);
+          break;
+        }
+      }
     },
-    [apiRef, colCount, logger, currentPage.rows],
+    [apiRef, colCount, currentPage, goToCell, goToHeader],
   );
 
   useGridApiEventHandler(apiRef, GridEvents.cellNavigationKeyDown, navigateCells);
```

Comment:
Since this file is being reworked.

```suggestion
  useGridApiEventHandler(apiRef, GridEvents.cellNavigationKeyDown, handleCellNavigationKeyDown);
```

Same for `GridEvents.columnHeaderNavigationKeyDown`.

Reply:
Better indeed, done !